transforms/sort.go

```diff
@@ -54,3 +56,75 @@ func Sort(less func(i, j optimus.Row) (bool, error)) optimus.TransformFunc {
 func StableSort(less func(i, j optimus.Row) (bool, error)) optimus.TransformFunc {
 	return sorter(sort.Stable, less)
 }
+
+type compressedRow struct {
+	key  string
+	blob []byte
+}
+
+type compressedSorter struct {
+	rows []compressedRow
+	err  error
+}
+
+func (r *compressedSorter) Len() int           { return len(r.rows) }
+func (r *compressedSorter) Less(i, j int) bool { return r.rows[i].key < r.rows[j].key }
+func (r *compressedSorter) Swap(i, j int)      { r.rows[i], r.rows[j] = r.rows[j], r.rows[i] }
+
+func compressedSort(sorter func(sort.Interface), getKey func(optimus.Row) (string, error)) optimus.TransformFunc {
+	return func(in <-chan optimus.Row, out chan<- optimus.Row) error {
+		rows := &compressedSorter{rows: []compressedRow{}}
+		for row := range in {
+			// obtain the key for the row and retain the key + blob while sorting
+			key, err := getKey(row)
+			if err != nil {
+				return err
+			}
+
+			rowBlob, err := json.Marshal(row)
+			if err != nil {
+				return err
+			}
+
+			rows.rows = append(rows.rows, compressedRow{
+				key:  key,
+				blob: rowBlob,
+			})
+		}
+
+		sorter(rows)
+
+		if rows.err != nil {
+			return rows.err
+		}
+		for _, row := range rows.rows {
+			var r optimus.Row
+			if err := json.Unmarshal(row.blob, &r); err != nil {
+				return err
+			}
+			out <- r
+		}
+		return nil
+	}
+}
+
+// StableCompressedSort takes in a function that reports whether the row i should sort before row j.
+// It outputs the rows in stably sorted order.
+func StableCompressedSort(getKey func(optimus.Row) (string, error)) optimus.TransformFunc {
+	return compressedSort(sort.Stable, getKey)
+}
+
+// GetKey is utility function for Optimus' StableCompressedSort functionality.
+func GetKey(key string) func(optimus.Row) (string, error) {
```

Comment:
It looks like this serves the same purpose as the RowIdentifier/KeyIdentifier functions that other transforms already use: https://godoc.org/gopkg.in/Clever/optimus.v3/transforms#RowIdentifier

Reply:
Getting an `interface{}` is fine for equality comparisons, how would you propose we do sorting? I'm not a fan of adding a type switch to handle various data types and that still wouldn't handle custom types/sort mechanisms w/o taking a `sort less(a, b interface{}) (bool, error)`.
Maybe I'm just biased because our use cases only need to work with string keys but these alternatives don't benefit us and lessen the compile-time checking we get.

Reply:
Unless I'm missing something, you don't get any more or less compile-time checks. You could error out if the identifier isn't a string - this function is still doing a run-time type assertion and using it to return a string, and erroring otherwise.

The only difference with using `RowIdentifier` is that you'd be re-using an already existing type meant for this purpose, and that that type assertion would live inside the transform instead of in this function.

Reply:
I kind of agree with Alex here. If we wanted to change this in the future, then we'd be in trouble. I'd rather keep the API consistent with the other API since it's an open source repo

Reply:
I think this is a mistake but if you both feel that way I'll do it ¯\\_(ツ)_/¯